web/auth/twofactor.go

```diff
@@ -0,0 +1,147 @@
+		return c.JSON(http.StatusBadRequest, "Missing twoFactorToken")
+	}
+
+	twoFactorToken := []byte(twoFactorTokenParam)
+
+	longRunSession, err := strconv.ParseBool(c.FormValue("long-run-session"))
+	if err != nil {
+		longRunSession = true
+	}
+
+	trustedDeviceCheckBoxParam := c.QueryParams().Get("trusted_device_checkbox")
+	if trustedDeviceCheckBoxParam != "" {
+		if b, err := strconv.ParseBool(trustedDeviceCheckBoxParam); err == nil {
+			trustedDeviceCheckBox = b
+		}
+	}
+
+	return renderTwoFactorForm(c, inst, http.StatusOK, credsError, redirect, twoFactorToken, longRunSession, trustedDeviceCheckBox)
+}
+
+// twoFactor handles a the twoFactor POST request
+func twoFactor(c echo.Context) error {
+	wantsJSON := c.Request().Header.Get(echo.HeaderAccept) == echo.MIMEApplicationJSON
+
+	inst := middlewares.GetInstance(c)
+	redirect, err := checkRedirectParam(c, inst.DefaultRedirection())
+	if err != nil {
+		return err
+	}
+
+	trustedDeviceCheckBox := true
+	trustedDeviceCheckBoxParam := c.QueryParams().Get("trusted_device_checkbox")
+	if trustedDeviceCheckBoxParam != "" {
+		if b, err := strconv.ParseBool(trustedDeviceCheckBoxParam); err != nil {
+			trustedDeviceCheckBox = b
+		}
+	}
+
+	// Retreiving data from request
+	token := []byte(c.FormValue("two-factor-token"))
+	passcode := c.FormValue("two-factor-passcode")
+	generateTrustedDeviceToken, _ := strconv.ParseBool(c.FormValue("two-factor-generate-trusted-device-token"))
+	longRunSession, _ := strconv.ParseBool(c.FormValue("long-run-session"))
+
+	correctPasscode := inst.ValidateTwoFactorPasscode(token, passcode)
+
+	// Handle 2FA failed
+	if !correctPasscode {
+		errorMessage := inst.Translate(TwoFactorErrorKey)
+		errCheckRateLimit := limits.CheckRateLimit(inst, limits.TwoFactorType)
+		if errCheckRateLimit == limits.ErrRateLimitExceeded {
+			if err := TwoFactorRateExceeded(inst); err != nil {
+				inst.Logger().WithField("nspace", "auth").Warning(err)
+				errorMessage = inst.Translate(TwoFactorExceededErrorKey)
+			}
+		}
+		// Renders either the passcode page or a JSON message
+		if wantsJSON {
+			return c.JSON(http.StatusUnauthorized, echo.Map{
+				"error": errorMessage,
+			})
+		}
+		return renderTwoFactorForm(c, inst, http.StatusUnauthorized, errorMessage, redirect, token, longRunSession, trustedDeviceCheckBox)
+	}
+
+	// Generates a new session
+	sessionID, err := newSession(c, inst, redirect, longRunSession)
+	if err != nil {
+		return err
+	}
+	// Check if the user trusts its device
+	var generatedTrustedDeviceToken []byte
+	if generateTrustedDeviceToken {
+		generatedTrustedDeviceToken, _ = inst.GenerateTwoFactorTrustedDeviceSecret(c.Request())
+	}
+	if wantsJSON {
+		result := echo.Map{"redirect": redirect.String()}
+		if len(generatedTrustedDeviceToken) > 0 {
+			result["two_factor_trusted_device_token"] = string(generatedTrustedDeviceToken)
+		}
+		return c.JSON(http.StatusOK, result)
+	}
+
+	redirect = AddCodeToRedirect(redirect, inst.ContextualDomain(), sessionID)
```

Comment:
Can you remember me what is the purpose of `AddCodeToRedirect`?

Reply:
Of what I understand, a generated `code` is added to the redirect. Its purpose is to be exchanged between main domain and subdomains (like apps) with a cookie.

This process allows to "transfer" a cookie for an instance configured with flat subdomains